integration/connector_test.go

```diff
@@ -0,0 +1,470 @@
+// Copyright (c) 2022 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package integration
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/eclipse/ditto-clients-golang"
+	"github.com/eclipse/ditto-clients-golang/model"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/eclipse/ditto-clients-golang/protocol/things"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/net/websocket"
+)
+
+type testConfig struct {
+	Broker                   string `def:"tcp://localhost:1883"`
+	MqttQuiesceMs            int    `def:"500"`
+	MqttAcknowledgeTimeoutMs int    `def:"3000"`
+
+	DittoAddress string
+
+	DittoUser     string `def:"ditto"`
+	DittoPassword string `def:"ditto"`
+
+	EventTimeoutMs  int `def:"30000"`
+	StatusTimeoutMs int `def:"10000"`
+
+	TimeDeltaMs int `def:"5000"`
+}
+
+type thingConfig struct {
+	DeviceID string `json:"deviceId"`
+	TenantID string `json:"tenantId"`
+	PolicyID string `json:"policyId"`
+}
+
+type commandPayload struct {
+	Topic   string                 `json:"topic"`
+	Headers map[string]interface{} `json:"headers"`
+	Path    string                 `json:"path"`
+	Value   string                 `json:"value"`
+}
+
+type commandResponsePayload struct {
+	commandPayload
+	Status int `json:"status"`
+}
+
+type ConnectorSuite struct {
+	suite.Suite
+
+	mqttClient  mqtt.Client
+	dittoClient *ditto.Client
+
+	cfg      *testConfig
+	thingCfg *thingConfig
+
+	thingURL   string
+	featureURL string
+}
+
+const (
+	envVariablesPrefix    = "SCT"
+	featureID             = "ConnectorTestFeature"
+	propertyName          = "testProperty"
+	commandName           = "testCommand"
+	commandResponseFormat = "response[%s]"
+)
+
+func (suite *ConnectorSuite) SetupSuite() {
+	cfg := &testConfig{}
+
+	suite.T().Log(getConfigHelp(*cfg, envVariablesPrefix))
+
+	if err := initConfigFromEnv(cfg, envVariablesPrefix); err != nil {
+		suite.T().Skip(err)
```

Comment:
If run with `go test` in this directory, this gets an error:
`Error initializing from config: env variable SCT_DITTO_ADDRESS not set`
and skips the test, declaring that it's successful.
I think we should fail the test instead.

Reply:
I beg to disagree. The test could be run as part of a set of tests - e.g. all suite connector tests - in an environment where cloud infrastructure is not available. It makes no sense to fail in that scenario.

More generally, this is indication of missing test precondition, not a bug in the tested functionality.

Reply:
If we automate this test in a properly setup environment, how do we distinguish between "preconditions failed" vs "test run successfully"? 

If the test fails on preconditions setup, then we can simply skip running it in a non-proper environment.

As it is, it can silently get skipped in an actual test, in case we got the environment wrong.

Reply:
> If we automate this test in a properly setup environment, how do we distinguish between "preconditions failed" vs "test run successfully"?

In the first case - the particular test will have `SKIP` status in the final report, in the second case it will have `PASS` status.

>As it is, it can silently get skipped in an actual test, in case we got the environment wrong.

It is not skipped silently - as you mentioned, a message denoting exactly which environmental variable is missing is displayed.

If your objection is, that the message can be misleading, because it contains the word "error", I can agree to that. Maybe a more suitable message would be `Could not initialize config - env variable SCT_DITTO_ADDRESS  not set`

Reply:
I still think that this should be a fail because right now, on a specific missing env variable, the test isn't run at all but the reported status is success (return code of 0). Otherwise, invoking bash scripts should check the output for the SKIP word, which is brittle.

Reply:
Could the new tests just use build tags? i.e. `go test` would *skip* these integration tests by default, but `go test --tags=integration` would only run the integration tests, failing hard if the environment setup isn't available.

Reply:
Done. We used the tag e2einttest, because integration is already being used for something else in the Kanto codebase.

Reply:
We have decided to use the tag `integration`.